### PR TITLE
DON-490 implemented new simple tightened email validation approach

### PR DESCRIPTION
**Pull request by alihejazi**

### CODE REVIEW

File: src/app/validators/validate-email.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+export function ValidateEmail(c: FormControl) {
+  // RegExp loosely based on SF charity onboarding one with some improvements for readability and length
+  // check on final a-z token. We found Angular's `Validators.email` too permissive.
+  const EMAIL_REGEXP = new RegExp('[a-z0-9.!#$%&*/=?^_+-`{|}~\'_%+-]+@[a-z0-9.-]+\\.[a-z]{2,10}$', 'i');
+
+  // Valid emails MUST pass regexp check and not include semi-colon(s) before the @ sign.
+  // The semi-colon check is because the regexp above can sometimes miss some invalid edge cases.
+  // See DON-490 for more information.
+  const isValid: boolean = EMAIL_REGEXP.test(c.value) && (!c.value.split('@')[0].includes(';'));
```
Comment (NoelLH) [inline src/app/validators/validate-email.ts:11]: @alihejazi  why not just change the regexp? This is adding a lot of unnecessary complexity to work out what's happening. And why do we need a whole new class and custom regexp execution logic for this when the build in `pattern(...)` validator is available?

**Review by NoelLH — CHANGES_REQUESTED** [review]
I'm not sure I understand why this is making such a big change instead of just to the pattern @alihejazi. The validation unit tests are also failing.

File: src/app/donation-start/donation-start.component.ts [hunk 2]
```diff
@@ -197,9 +197,13 @@ export class DonationStartComponent implements AfterContentChecked, AfterContent
...
-          // RegExp loosely based on SF charity onboarding one with some improvements for readability and length
-          // check on final a-z token. We found Angular's `Validators.email` too permissive.
-          Validators.pattern(new RegExp('[a-z0-9.!#$%&*/=?^_+-`{|}~\'_%+-]+@[a-z0-9.-]+\\.[a-z]{2,10}$', 'i')),
+          Validators.email,
+          // Validators.email regexp above rejects most invalid emails but has a few edge-cases slip through.
+          // For example, it allows emails ending with numbers like hello@thebiggive.org.uk.123
+          // However, this is invalid, so we need some tighter validation. The additional regexp below
+          // ensures that emails finish with format like .org.uk or .abc.abc or .io or .com or .whatever
+          // as long as there are no numbers. See ticked DON-490.
+          Validators.pattern(new RegExp('(\.[a-z]{2,})+$', 'i')),
```
Comment (NoelLH) [inline src/app/donation-start/donation-start.component.ts:206]: @alihejazi as flagged by GitHub, the `.` here is not doing what you think. By having any number of _anything then 2 letter_ repeats at the end I expect this opens up edge cases that are not intended.

**Review by NoelLH — CHANGES_REQUESTED** [review]
@alihejazi I'm not too sure about this approach in general – regular expressions are relatively heavy/slow operations and `.email` will be using one internally. And stacking two this way is likely to make debugging remaining issues hard.

I think there are also some specific problems with the new pattern, detailed inline.

File: src/app/donation-start/donation-start.component.ts [hunk 3]
```diff
@@ -156,6 +156,8 @@ export class DonationStartComponent implements AfterContentChecked, AfterContent
     this.campaign = this.route.snapshot.data.campaign;
     this.setCampaignBasedVars();
 
+    const emailRegex : RegExp = /^(?=.{1,254}$)(?=[a-zA-Z0-9+=_.-]{1,64}@)[a-zA-Z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-zA-Z0-9!#$%&'*+/=?^_`{|}~-]+)*@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(\.[a-zA-Z]{2,})+$/;
```
[the remark this answers is not in the document]
Comment (alihejazi) [reply src/app/donation-start/donation-start.component.ts:159]: @NoelLH one of the disadvantages of keeping the original `(?=.{1,64}@)` is that it will mean the email _can_ have special characters in the username section. For example, `a&^^.!!!!$%li@thebiggive.org.uk` would be valid. If you're okay with this then sure I'll proceed to change it back to `(?=.{1,64}@)`.

Apologies I didn't get what you mean with your last sentence. Do you mean to clarify on Jira that the `\.` is no longer an issue due to using the `/^.../` syntax?

[the remark this answers is not in the document]
Comment (NoelLH) [reply src/app/donation-start/donation-start.component.ts:159]: @alihejazi I think that 'weird' characters in the username part are fine given they are allowed by the specs and most people won't use this anyway. If we are keeping support for a username part at all, which I think we might as well, then I think it makes sense to keep the pattern simple and the constraints as similar as possible to other implementations rather than inventing our own additional ones.

To your second point, I just meant that I was confused that your pattern after _"It’s now been changed to this:"_ does not match the PR – in terms of the dot escape (which is part of the pattern and not just a string escape) but also in other larger ways. This makes it quite confusing to tie the reasoning for the changes to the actual code being proposed and therefore to review it effectively.

[the remark this answers is not in the document]
Comment (alihejazi) [reply src/app/donation-start/donation-start.component.ts:159]: @NoelLH Ok, I will change the username part back to the original `(?=.{1,64}@)`

Regarding your second point, I just realised the reason why my comment on Jira doesn't match this PR was because Jira rendered out some characters that are used for styling. For example the regex has two pairs of _, which have been rendered out as they're used for italics, and had at least two occurrences of ` too. I've just changed the format to make the regexp show in code blocks, not embedded code within the comments.
